analyzers/src/SonarAnalyzer.VisualBasic/Rules/UnusedStringBuilder.cs

```diff
@@ -0,0 +1,65 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using StyleCop.Analyzers.Lightup;
+
+namespace SonarAnalyzer.Rules.VisualBasic;
+
+[DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+public sealed class UnusedStringBuilder : UnusedStringBuilderBase<SyntaxKind, VariableDeclaratorSyntax, IdentifierNameSyntax>
+{
+    private static readonly SyntaxKind[] SkipChildren = { };
+
+    protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
+
+    protected override SyntaxNode GetScope(VariableDeclaratorSyntax declarator) =>
+        declarator is { Parent: LocalDeclarationStatementSyntax { Parent: { } block } }
+        ? block
+        : null;
+
+    protected override ILocalSymbol RetrieveStringBuilderObject(SemanticModel semanticModel, VariableDeclaratorSyntax declarator) =>
+        declarator is
+        {
+            Parent: LocalDeclarationStatementSyntax,
+            Initializer.Value: ObjectCreationExpressionSyntax { } objectCreation,
+            Names: { Count: 1 } names, // Must be 1, because otherwise "BC30671: Explicit initialization is not permitted with multiple variables declared with a single type specifier." is raised.
+        }
+        && objectCreation.Type.IsKnownType(KnownType.System_Text_StringBuilder, semanticModel)
+            ? semanticModel.GetDeclaredSymbol(names[0]) as ILocalSymbol
+            : null;
+
+    protected override bool IsStringBuilderRead(SemanticModel model, ILocalSymbol symbol, SyntaxNode node) =>
+        node switch
+        {
+            InvocationExpressionSyntax invocation =>
+                (IsAccessInvocation(invocation.GetName()) && IsSameReference(model, symbol, invocation.Expression))
+                || (IsSameReference(model, symbol, invocation.Expression) && model.GetOperation(invocation).Kind is OperationKindEx.PropertyReference),
```

Comment:
You are right about the syntax check: There is no way to detect whether the invocation is a method or an indexer. You can use [IPropertySymbol.IsIndexer](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.ipropertysymbol.isindexer?view=roslyn-dotnet-4.3.0) instead of IOperation though.

Reply:
You are right! Nice find 👍